### PR TITLE
Backport L.Map.CompassBearing from rotate branch

### PR DESCRIPTION
**Pull request by fnicollet**
Fixes https://github.com/hyperknot/Leaflet/issues/19

### CODE REVIEW

File: src/map/handler/Map.CompassBearing.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/*
+ * L.Map.CompassBearing will rotate the map according to a smartphone's compass.
+ */
+
+L.Map.CompassBearing = L.Handler.extend({
+
+	initialize: function(map) {
+
```
Comment (hyperknot) [inline src/map/handler/Map.CompassBearing.js]: We should check if map is rotation enabled and only initialize if it is.

**Comment by fnicollet** [comment]
Added a check on initialize

**Comment by hyperknot** [comment]
My problem is that the compass probing now runs on all inits, even if the user doesn't need it. Any reason why using initialize instead of a simple addHooks?

**Comment by fnicollet** [comment]
Not sure what you mean :) Can you do the edit on your side?

Fabien

On Tue, Jun 28, 2016 at 4:04 AM, Zsolt Ero <notifications@github.com> wrote:

> My problem is that the compass probing now runs on all inits, even if the
> user doesn't need it. Any reason why using initialize instead of a simple
> addHooks?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/pull/20#issuecomment-228928710>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABLtoXqotd6ZOqck4SnZqhvZ9i2QcmnDks5qQIFHgaJpZM4I5Q_d>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Comment by hyperknot** [comment]
I see, this is a port from Ivan's code. So the problem is that initialize and probing always runs, on all rotation enabled maps. It should only run if it is actually enabled.

**Comment by fnicollet** [comment]
Hum, if I take one from leaflet's core:
https://github.com/Leaflet/Leaflet/blob/master/src/map/handler/Map.DoubleClickZoom.js
There is no special check. Shouldn't "addInitHook" check whether
"compassBearing" is falsey already?

On Tue, Jun 28, 2016 at 2:21 PM, Zsolt Ero <notifications@github.com> wrote:

> I see, this is a port from Ivan's code. So the problem is that initialize
> and probing always runs, on all rotation enabled maps. It should only run
> if it is actually enabled.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/pull/20#issuecomment-229032911>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABLtoWkuCrT3HuRAvXace3m_pEfQocpFks5qQRHXgaJpZM4I5Q_d>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Comment by hyperknot** [comment]
@fnicollet I looked at it and I think DoubleClickZoom is the perfect handler to base CompassBearing on.

Have a look at how Handler is implemented: https://github.com/Leaflet/Leaflet/blob/master/src/core/Handler.js

So there is a map option, which together with addInitHook adds a handler, whose addHooks and removeHooks method is called. All we need to do is to make the probing logic simpler and put it into addHooks, so it only runs when added.

**Comment by fnicollet** [comment]
I see what you mean now, this file need a bit of refactoring then to move
the logic to addHooks / removeHooks

I'll try to give it a go, I won't be mad if you do it before me ahah :)

On Tue, Jun 28, 2016 at 3:04 PM, Zsolt Ero <notifications@github.com> wrote:

> @fnicollet <https://github.com/fnicollet> I looked at it and I think
> DoubleClickZoom is the perfect handler to base CompassBearing on.
>
> Have a look at how Handler is implemented:
> https://github.com/Leaflet/Leaflet/blob/master/src/core/Handler.js
>
> So there is a map option, which together with addInitHook adds a handler,
> whose addHooks and removeHooks method is called. All we need to do is to
> make the probing logic simpler and put it into addHooks, so it only runs
> when added.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/pull/20#issuecomment-229042344>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABLtobpnpSupxHeVyd5bgkKvFxA8anecks5qQRvDgaJpZM4I5Q_d>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Comment by hyperknot** [comment]
You'll probably use this part more than me, so I'll let you come up with the solution you like the most.

**Comment by fnicollet** [comment]
Following your advice, I completely changed the class to make it more simple (probing wasn't really necessary) and compatible with addHooks/removeHooks

**Comment by hyperknot** [comment]
I merged it by rebasing manually and also added a check for event.alpha !== null

**Comment by fnicollet** [comment]
Great, thanks :)

On Tue, Jun 28, 2016 at 11:59 PM, Zsolt Ero <notifications@github.com>
wrote:

> Closed #20 <https://github.com/hyperknot/Leaflet/pull/20>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperknot/Leaflet/pull/20#event-707007531>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABLtoUMYWuXCynM9bI03LhFLU8FXCH-yks5qQZlegaJpZM4I5Q_d>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Pull request closed without merging**